lib/membrane_rtc_engine/endpoints/rtsp_endpoint.ex

```diff
@@ -77,37 +132,145 @@ if Enum.all?(
+          :rtp,
+          _ctx,
+          _state
+        ) do
+      raise("Received unexpected, second RTP stream: #{inspect(msg)}")
+    end
+
+    @impl true
+    def handle_child_notification({:estimation, _data}, {:track_sender, _track_id}, _ctx, state) do
+      {[], state}
     end
 
     @impl true
-    def handle_child_notification(notification, element, _context, state) do
+    def handle_child_notification(notification, element, _ctx, state) do
       Membrane.Logger.warn(
         "Unexpected notification from `#{inspect(element)}`: #{inspect(notification)}. Ignoring."
       )
 
       {[], state}
     end
 
+    @impl true
+    def handle_info({:rtsp_setup_complete, options}, ctx, state) do
+      Membrane.Logger.debug("Endpoint received source options: #{inspect(options)}")
+
+      {:ok, fmtp} = ExSDP.Attribute.FMTP.parse(options[:fmtp])
+      rtpmap = options[:rtpmap]
+
+      encoding = rtpmap.encoding |> String.to_atom()
+
+      if encoding != :H264 do
+        raise("RTSP setup returned unsupported stream encoding: #{inspect(encoding)}")
+      end
+
+      {:endpoint, endpoint_id} = ctx.name
+
+      track =
+        Track.new(
+          :video,
+          Track.stream_id(),
+          endpoint_id,
+          encoding,
+          rtpmap.clock_rate,
+          fmtp,
+          ctx: %{rtpmap: rtpmap}
+        )
+
+      Membrane.Logger.debug("Publishing new WebRTC track: #{inspect(track)}")
+      actions = [notify_parent: {:publish, {:new_tracks, [track]}}]
+      state = %{state | track: track}
+
+      if is_nil(state.stream_params) do
+        Membrane.Logger.debug(
+          "Track will be marked as ready once an RTP stream has been received"
+        )
+
+        {actions ++ create_nat_binding_actions(state, options[:server_port]), state}
+      else
+        {actions ++ ready_track_actions(state), state}
+      end
+    end
+
     @impl true
     def handle_info(info, _ctx, state) do
       Membrane.Logger.warn("Unexpected info: #{inspect(info)}. Ignoring.")
       {[], state}
     end
+
+    defp ready_track_actions(state) do
+      if state.track.ctx.rtpmap.payload_type != state.stream_params.fmt do
+        raise("Payload type mismatch between RTP mapping and received stream")
+      end
+
+      [notify_parent: {:track_ready, state.track.id, :high, state.track.encoding}]
+    end
+
+    defp create_nat_binding_actions(state, server_port) do
+      Membrane.Logger.debug(
+        "Probing server port #{inspect(server_port)} to attempt to create NAT binding..."
+      )
+
+      # FIXME: The following assumes IPv4, which may not always be the case
+      # Also, this is a bit too low-level for the endpoint,
+      # but there really is no other element which could handle it at the moment...
+      {:ok, address} =
+        URI.parse(state.source_uri)
+        |> Map.get(:host)
+        |> to_charlist()
+        |> :inet.getaddr(:inet)
+
+      [notify_child: {:udp_source, {:probe, address, server_port}}]
```

Comment:
This is the current attempt to create client-side NAT binding

It requires the following callback to be implemented in `Membrane.UDP.Source`:
```elixir
  @impl true
  def handle_parent_notification({:probe, ip, port}, _ctx, state) do
    Socket.send(%Socket{ip_address: ip, port_no: port}, state.local_socket, <<>>)
    {[], state}
  end
```

Any and all comments about this solution (flawed and hack-y, no doubt) are welcome here

Reply:
Good idea. Will look into it